Pydantic V2 blog (#4218)
samuelcolvin/pydantic: docs/blog/pydantic-v2.md

I want it to be as good as possible.

While I'm on the subject of why, how and my odd sabbatical: if you work for a large company who use pydantic a lot,
you should encourage the company to **sponsor me a meaningful amount**, 
like [Salesforce did](https://twitter.com/samuel_colvin/status/1501288247670063104).
This is not charity, recruitment or marketing - the argument should be about how much the company will save if
pydantic is 10x faster, more stable and more powerful - it would be worth paying me 10% of that to make it happen.

The plan is to have pydantic V2 released within 3 months of full time work 
(again, that'll be sooner if I can continue to work on it full time).

Before pydantic V2 can be released, we need to released pydantic v1.10 - there are lots of changes in the main
branch of pydantic contributed by the community, it's only fair to provide a release including those changes,
many of them will remain unchanged for V2, the rest will act as a requirement to make sure pydantic V2 includes
the capabilities they implemented.

The basic road map for me is as follows:

> 1. implement a few more critical features in pydantic-core

Reviewer: could you be explicit here?

4. release pydantic V1.10
5. delete all stale PRs which didn't make it into V1.10, apologise profusely to their authors who put their valuable
  time into pydantic only to have their PRs closed :pray:
6. change the main branch of pydantic to target v2
7. start tearing pydantic code apart and see how many existing tests can be made to pass
8. rinse, repeat
9. release pydantic V2 :tada:

# Introduction

Pydantic began life as an experiment in some code in a long dead project.
I ended up making the code into a package and releasing it.
It got a bit of attention on hacker news when it was first released, but started to get really popular when
Sebastian Ramirez used it in FastAPI.
Since then the package and its usage have grown enormously.
The core logic however has remained relatively unchanged since the initial experiment.
It's old, it smells, it needs to be rebuilt.
The release of version 2 is an opportunity to rebuild pydantic and correct many things that don't make sense.

Much of the work on V2 is already done, but there's still a lot to do.
Now seems a good opportunity to explain what V2 is going to look like and get feedback from uses.

> ## Headlines

Reviewer: I think this section could be clearer, it kind of bleeds into the next section of proposed changes and could be more punchy.

Author: rewritten.

> For good and bad, here are some of the biggest changes expected in V2.

Author: Because some downsides are inevitable, at least in some people's opinions.

> For good and bad, here are some of the biggest changes expected in V2.

Author: But might re-write.

The core validation logic of pydantic V2 will be performed by a separate package 
[pydantic-core](https://github.com/samuelcolvin/pydantic-core) which I've been building over the last few months.

*pydantic-core* is written in Rust using the excellent [pyo3](https://pyo3.rs/) library which provides rust bindings
for python.

**Note:** the python interface to pydantic shouldn't change as a result of using pydantic-core, instead
pydantic will use type annotations to build a schema for pydantic-core to use.

pydantic-core is usable now, albeit with a fairly unintuitive API, if you're interested, please give it a try.

pydantic-core provides validators for all common data types, 
[see a list here](https://github.com/samuelcolvin/pydantic-core/blob/main/pydantic_core/_types.py#L291).
Other, less commonly used data types will be supported via validator functions.

### Performance :smile:

> As a result of the move to rust for the validation logic 

Reviewer: Did. you move pydantic-core to rust because of performance reasons? Up to here you haven't shared the reasoning for that change. It might be helpful to say "We desire 10x performance, candidates were A, B.  I chose rust because performance, Y, Z."

Author: done.

pydantic V2 is between 4x and 50x faster than pydantic V1.X.

In general, pydantic V2 is about 17x faster than V1.X when validating a representative model containing a range
of common fields.

### Strict Mode :smile:

People have long complained about pydantic preference for coercing data instead of throwing an error.
E.g. input to an `int` field could be `123` or the string `"123"` which would be converted to `123`.

pydantic-core comes with "strict mode" built in. With this only the exact data type is allowed, e.g. passing
`"123"` to an `int` field would result in a validation error.

Strictness can be defined on a per-field basis, or whole model.

#### IsInstance checks :smile:

Strict mode also means it makes sense to provide an `is_instance` method on validators which effectively run
validation then throw away the result while avoiding the (admittedly small) overhead of creating and raising
and error or returning the validation result.

### Formalised Conversion Table :smile:

As well as complaints about coercion, another (legitimate) complaint was inconsistency around data conversion.

> In pydantic V2, the following principle will govern when data should be converted in "lax mode" (`strict=False`):

Reviewer: If you're not going to use this terminology everywhere I'd suggest just calling it "normal" mode, or "non strict mode", if there are only two modes - normal / default, and strict.

Author: it's called lax throughout the code base. I'll stick with lax vs. strict.

The only place where I know this will cause problems is Raspberry Pi, which is a bit of 
[mess](https://github.com/piwheels/packages/issues/254) when it comes to packages written in rust for python.
Effectively, until that's fixed you'll likely have to install pydantic-core with 
`pip install -i https://pypi.org/simple/ pydantic-core`.

### Pydantic becomes a pure python package :confused:

Pydantic v1.X is a pure python code base but is compiled with cython to provide some performance improvements.
Since the "hot" code is moved to pydantic-core, pydantic itself can go back to being a pure python package.

This should significantly reduce the size of the pydantic package and make unit tests of pydantic much faster.
In addition, some constraints on pydantic code can be removed once it no-longer has to be compilable with cython.

Some pieces of edge logic could get a little slower as they're no longer compiled.

## Other Improvements :smile:

* Recursive models
* Documentation examples you can edit and run

TODO

## Removed Features :neutral_face:

* `__root__` models

TODO

## Conversion Table

> TODO

Reviewer: Consider adding an ending section / wrapup. The blog just trails off with these features so either add a short sum-up at the end or have some conclusions *before* this list of features and proposed changes and say see below for the full list. 

Given folks will have lots of opinions about these changes, it could be worth getting ahead of that and explicitly stating how you would like folks to contribute thoughts, opinions etc. E.g. an issue thread?

Author: done.

Other, less commonly used data types will be supported via validator functions.

### Performance :thumbsup:

As a result of the move to rust for the validation logic 
(and significant improvements in how validation objects are structured) pydantic V2 will be significantly faster
than pydantic V1.

Looking at the pydantic-core [benchmarks](https://github.com/samuelcolvin/pydantic-core/tree/main/tests/benchmarks)
today, pydantic V2 is between 4x and 50x faster than pydantic V1.9.1.

In general, pydantic V2 is about 17x faster than V1 when validating a model containing a range
of common fields.

### Strict Mode :thumbsup:

People have long complained about pydantic for coercing data instead of throwing an error.
E.g. input to an `int` field could be `123` or the string `"123"` which would be converted to `123`.

pydantic-core comes with "strict mode" built in. With this only the exact data type is allowed, e.g. passing
`"123"` to an `int` field would result in a validation error.

This will allow pydantic V2 to offer a `strict` switch which can be set on either a model or a field.

> #### `IsInstance` checks :thumbsup:

Reviewer 2: That will be very nice even for tests (this + dirty-equals should make shape testing super easy)

!!! note
    Pydantic has always had special support for JSON, that is not going to change.

    While in theory other formats
    could be specifically supported, the overheads are significant and I don't think there's another format that's
    used widely enough to be worth specific logic. Other formats can be parsed to python then validated, similarly
    when serialising, data can be exported to a python object, then serialised, 
    see [below](#improvements-to-dumpingserializationexport).

### Validation without a Model :thumbsup:

In pydantic V1 the core of all validation was a pydantic model, this led to significant overheads and complexity
when the output data type was not a model.

pydantic-core operates on a tree of validators with no "model" type required at the base of the tree.
It can therefore validate a single `string` or `datetime` value, a `TypeDict` or `Model` equally easily.

This feature will provide significant addition performance improvements in scenarios like:

* adding validation to `dataclasses`
* validating URL arguments, query strings, headers, etc. in FastAPI
* adding validation to `TypedDict`

> * function argument validation

Reviewer 2: I would have added "adding validation to your custom classes, decorators...". Thing that has been tried many times with pydantic models with custom metaclasses or whatever (see issues)

* `._model_validator` attribute holding the internal pydantic `SchemaValidator`
* `.model_fields` (currently `.__fields__`) - although the format will have to change a lot, might be an alias of
  `._model_validator.schema`
* `ModelConfig` (currently `Config`) - configuration class for models

The following methods will be removed:

* `.parse_file()`
* `.parse_raw()`
* `.from_orm()` - the functionality has been configured as a `Config` property call `from_attributes` which can be set
  either on `ModelConfig` or on a specific `ModelClass` or `TypedDict` field

### Strict API & API documentation :thumbsup:

When preparing a pydantic V2, we'll make a strict distinction between the public API and private functions & classes.
Private objects clearly identified as private via `_internal` sub package to discourage use.

The public API will have API documentation. I've recently been working with the wonderful
[mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) for both 
[dirty-equals](https://dirty-equals.helpmanual.io/) and
[watchfiles](https:://watchfiles.helpmanual.io/) documentation. I intend to use `mkdocstrings` to generate complete
API documentation.

> This wouldn't replace the current example-based somewhat informal documentation style byt instead will augment it.

Reviewer 2: ```suggestion
This wouldn't replace the current example-based somewhat informal documentation style but instead will augment it.
```

### Error descriptions :thumbsup:

The way line errors (the individual errors with a `ValidationError`) are built has become much more sophisticated
in pydantic-core.

> There's a well=defined 

Reviewer 2: ```suggestion
There's a well-defined 
```

Author: fixed already 🤦

* **Windows**: `amd64` & `win32`
* **Web Assembly**: `wasm32` (pydantic-core is already compiled for wasm32 using emscripten and unit tests pass, 
  except where cpython itself has [problems](https://github.com/pyodide/pyodide/issues/2841))

Other binaries can be added provided they can be (cross-)compiled on github actions.
If no binary is available from PyPI, pydantic-core can be compiled from source if Rust stable is available.

The only place where I know this will cause problems is Raspberry Pi, which is a bit of 
[mess](https://github.com/piwheels/packages/issues/254) when it comes to packages written in rust for python.
Effectively, until that's fixed you'll likely have to install pydantic-core with 
`pip install -i https://pypi.org/simple/ pydantic-core`.

### Pydantic becomes a pure python package :neutral_face:

Pydantic V1.X is a pure python code base but is compiled with cython to provide some performance improvements.
Since the "hot" code is moved to pydantic-core, pydantic itself can go back to being a pure python package.

This should significantly reduce the size of the pydantic package and make unit tests of pydantic much faster.

> In addition, some constraints on pydantic code can be removed once it no-longer has to be compilable with cython.

Reviewer 2: It will also allow us to have proper generic constraints types and models

!!! note
    Pydantic has always had special support for JSON, that is not going to change.

    While in theory other formats
    could be specifically supported, the overheads are significant and I don't think there's another format that's
    used widely enough to be worth specific logic. Other formats can be parsed to python then validated, similarly
    when serialising, data can be exported to a python object, then serialised, 
    see [below](#improvements-to-dumpingserializationexport).

### Validation without a Model :thumbsup:

In pydantic V1 the core of all validation was a pydantic model, this led to significant overheads and complexity
when the output data type was not a model.

pydantic-core operates on a tree of validators with no "model" type required at the base of that tree.
It can therefore validate a single `string` or `datetime` value, a `TypeDict` or a `Model` equally easily.

This feature will provide significant addition performance improvements in scenarios like:

* adding validation to `dataclasses`
* validating URL arguments, query strings, headers, etc. in FastAPI
* adding validation to `TypedDict`
* function argument validation

> Basically anywhere were you don't care about a traditional model class instance.

Reviewer 3: ```suggestion
Basically anywhere where you don't care about a traditional model class instance.
```

the terms "parse" and "validate" for what pydantic does.

While pydantic is not simply a validation library (and I'm sure some would argue validation is not strictly what it does),
most people use the word **"validation"**.

It's time to stop fighting that, and use consistent names.

The word "parse" will no longer be used except when talking about JSON parsing, see
[model methods](#model-namespace-cleanup) above.

## Changes to custom field types :neutral_face:

Since the core structure of validators has changed from "a list of validators to call one after another" to
"a tree of validators which call each other", the 
[`__get_validators__`](https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__)
way of defining custom field types no longer makes sense.

Instead we'll look for the attribute `__pydantic_schema__` which must be a 
pydantic-core compliant schema for validating data to this field type (the `function`
item can be a string, if so a function of that name will be taken from the class, see `'validate'` below).

Here's an example of how a custom field type could be defined:

```py title="New custom field types"

> from pydantic import ValiationSchema

Reviewer 2: ValidationSchema

3. Release pydantic V1.10
4. Delete all stale PRs which didn't make it into V1.10, apologise profusely to their authors who put their valuable
   time into pydantic only to have their PRs closed :pray: 
   (and explain when and how they can rebase and recreate the PR)
5. Rename `master` to `main`, seems like a good time to do this
6. Change the main branch of pydantic to target V2
7. Start tearing pydantic code apart and see how many existing tests can be made to pass
8. Rinse, repeat
9. Release pydantic V2 :tada:

Plan is to have all this done by the end of October, definitely by the end of the year.

### Breaking Changes & Compatibility :pray:

While we'll do our best to avoid breaking changes, some things will break.

As per the [greatest pun in modern TV history](https://youtu.be/ezAlySFluEk).

> You can't make a Tomelette without breaking some Greggs.

Where possible, if breaking changes are unavoidable, we'll try to provide warnings or errors to make sure those 
changes are obvious to developers.

## Motivation & `pydantic-core`

Since pydantic's initial release, with the help of wonderful contributors 
[Eric Jolibois](https://github.com/PrettyWood),

> [Sebastián Ramírez](https://github.com/tiangolo),

Author: No, thank you for your contributions.

(and significant improvements in how validation objects are structured) pydantic V2 will be significantly faster
than pydantic V1.

Looking at the pydantic-core [benchmarks](https://github.com/samuelcolvin/pydantic-core/tree/main/tests/benchmarks)
today, pydantic V2 is between 4x and 50x faster than pydantic V1.9.1.

In general, pydantic V2 is about 17x faster than V1 when validating a model containing a range
of common fields.

### Strict Mode :thumbsup:

People have long complained about pydantic for coercing data instead of throwing an error.
E.g. input to an `int` field could be `123` or the string `"123"` which would be converted to `123`.

pydantic-core comes with "strict mode" built in. With this only the exact data type is allowed, e.g. passing
`"123"` to an `int` field would result in a validation error.

This will allow pydantic V2 to offer a `strict` switch which can be set on either a model or a field.

#### `is_instance` checks :thumbsup:

Strict mode also means it makes sense to provide an `is_instance` method on validators which effectively run
validation then throws away the result while avoiding the (admittedly small) overhead of creating and raising

> and error or returning the validation result.

Author: on second point https://github.com/samuelcolvin/pydantic-core/issues/155

> and error or returning the validation result.

Reviewer 2: @samuelcolvin Could it make sense to add a kwarg to choose between covariance and invariance?

> and error or returning the validation result.

Author: Can you give an example of what you mean?

> and error or returning the validation result.

Author: I don't think so, since we're just using pydantic's validation logic, we can use strict mode, but no more.

| `int`      | `b"1"`                  | :material-close:      | :material-close: | Error   |

(For the last case converting `bytes` to an `int` could reasonably mean `int(bytes_data.decode())` or 
`int.from_bytes(b'1', 'big/little')`, hence an error)

In addition to the general rule, we'll provide a conversion table which defines exactly what data will be allowed 
to which field types. See [the table below](#conversion-table) for a start on this.

### Built in JSON support :thumbsup:

pydantic-core can parse JSON directly into a model or output type, this both improves performance and avoids
issue with strictness - e.g. if you have a strict model with a `datetime` field, the input must be a 
`datetime` object, but clearly that makes no sense when parsing JSON which has no `datatime` type.
Same with `bytes` and many other types.

Pydantic V2 will therefore allow some conversion when validating JSON directly, even in strict mode 
(e.g. `ISO8601 string -> datetime`, `str -> bytes`) even though this would not be allowed when validating
a python object.

In future direct validation of JSON will also allow:

* parsing in a separate thread while starting validation in the main thread

> * line numbers from JSON to be included in the validation errors

Reviewer 4: Amazing! :tada:

!!! note
    Pydantic has always had special support for JSON, that is not going to change.

    While in theory other formats
    could be specifically supported, the overheads are significant and I don't think there's another format that's
    used widely enough to be worth specific logic. Other formats can be parsed to python then validated, similarly
    when serialising, data can be exported to a python object, then serialised, 
    see [below](#improvements-to-dumpingserializationexport).

### Validation without a Model :thumbsup:

In pydantic V1 the core of all validation was a pydantic model, this led to significant overheads and complexity
when the output data type was not a model.

pydantic-core operates on a tree of validators with no "model" type required at the base of that tree.
It can therefore validate a single `string` or `datetime` value, a `TypeDict` or a `Model` equally easily.

This feature will provide significant addition performance improvements in scenarios like:

* Adding validation to `dataclasses`

> * Validating URL arguments, query strings, headers, etc. in FastAPI

Reviewer 4: Awesome! Up to now, it's been the internal pydantic Field objects, awesome to have official first-class support for all these things.

* Adding validation to `TypedDict`
* Function argument validation

> * Adding validation to your custom classes, decorators...

Reviewer 4: Yessss! Type annotations FTW :rocket:

In effect - anywhere where you don't care about a traditional model class instance.

> We'll need to add standalone methods for generating json schema and dumping these objects to JSON etc.

Reviewer 4: Nitpick typo

```suggestion
We'll need to add standalone methods for generating JSON Schema and dumping these objects to JSON, etc.
```

> ### Required vs. Nullable Cleanup :thumbsup:

Reviewer 4: YESSSSS! :tada: :tada: :tada:

converting a model to a dict.

My plan is to move data export into pydantic-core, with that, one implementation can support all export modes without
compromising (and hopefully significantly improving) performance.

I see four different export/serialisation scenarios:

1. Extracting the field values of a model with no conversion, effectively `model.__dict__` but with the current filtering
   logic provided by `.dict()`
2. Extracting the field values of a model recursively (effectively what `.dict()` does now) - sub-models are converted to
   dicts, but other fields remain unchanged.
3. Extracting data and converting at the same time (e.g. to JSON compliant types)
4. Serialising data straight to JSON

I think all 4 modes can be supported in a single implementation, with a kind of "3.5" mode where a python function
is used to convert the data as the use wishes.

The current `include` and `exclude` logic is extremely complicated, but hopefully it won't be too hard to
translate it to Rust.

We should also add support for `validate_alias` and `dump_alias` as well as the standard `alias`
to allow for customising field keys.

> ### Model namespace cleanup :thumbsup:

Author: hadn't thought about that, good point.

  `.__model_validator__.schema`
* `.model_customise_schema()` - a function to customise the schema used to build the `pydantic_core.SchemaValidator`
  this provides a way to add or change validation without having to use a custom type
* `ModelConfig` (currently `Config`) - configuration class for models

The following methods will be removed:

* `.parse_file()`
* `.parse_raw()`
* `.from_orm()` - the functionality has been moved to config, see [other improvements](#other-improvements) below

### Strict API & API documentation :thumbsup:

When preparing for pydantic V2, we'll make a strict distinction between the public API and private functions & classes.
Private objects will be clearly identified as private via a `_internal` sub package to discourage use.

The public API will have API documentation. I've recently been working with the wonderful
[mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) package for both 
[dirty-equals](https://dirty-equals.helpmanual.io/) and
[watchfiles](https://watchfiles.helpmanual.io/) documentation. I intend to use `mkdocstrings` to generate complete
API documentation for V2.

> This wouldn't replace the current example-based somewhat informal documentation style but instead will augment it.

Reviewer 4: Nice! I've been asked for this recently in FastAPI. :sweat_smile: 

---

Somewhat irrelevant rant no one needs to read:

I would like to find a way to document parameters *inline*, using `Annotated`, instead of having to remember to keep docstrings in sync, also because docstrings are based on some custom format without editor help.

I think some type of parameter docs with `Annotated`, and having a way to annotate/decorate things (functions and classes) and mark them as discouraged/deprecated in a "standard" way would be quite helpful for Python. And if editors supported it, if it was some type of standard, it would help the experience of writing and maintaining code docs.

Author: I hadn't thought of `Annotated`, pycharm does a pretty good job of highlighting markdown in docstrings.

On annotating attributes, https://github.com/mkdocstrings/mkdocstrings/issues/446 might be of interest to you.

Have a play with the watchfiles codebase.

Reviewer 4: Nice, thank you!

### Error descriptions :thumbsup:

The way line errors (the individual errors within a `ValidationError`) are built has become much more sophisticated
in pydantic-core.

There's a well-defined 
[set of error codes and messages](https://github.com/samuelcolvin/pydantic-core/blob/main/src/errors/kinds.rs).

> More will be added when other type are validated via pure python validators in pydantic.

Reviewer 4: Microtypo

```suggestion
More will be added when other types are validated via pure python validators in pydantic.
```

I would like to add a dedicated section to the documentation with extra information for each type of error.

> This would be another key in a line error: `documentation`, which would link to the appropriate section in the

Reviewer 4: This is very cool!

Author: great, pleased you like the idea.

  [problems](https://github.com/pyodide/pyodide/issues/2841))

Other binaries can be added provided they can be (cross-)compiled on github actions.
If no binary is available from PyPI, pydantic-core can be compiled from source if Rust stable is available.

The only place where I know this will cause problems is Raspberry Pi, which is a 
[mess](https://github.com/piwheels/packages/issues/254) when it comes to packages written in Rust for python.
Effectively, until that's fixed you'll likely have to install pydantic with 
`pip install -i https://pypi.org/simple/ pydantic`.

### Pydantic becomes a pure python package :thumbsup:

Pydantic V1.X is a pure python code base but is compiled with cython to provide some performance improvements.
Since the "hot" code is moved to pydantic-core, pydantic itself can go back to being a pure python package.

This should significantly reduce the size of the pydantic package and make unit tests of pydantic much faster.
In addition, some constraints on pydantic code can be removed once it no-longer has to be compilable with cython.

Some pieces of edge logic could get a little slower as they're no longer compiled.

> ### I'm dropping the word "parse" and just using "validate" :neutral_face:

Author: 😨 🔫

Partly due to the issues with the lack of strict mode,
I've gone back and forth between using the terms "parse" and "validate" for what pydantic does.

While pydantic is not simply a validation library (and I'm sure some would argue validation is not strictly what it does),
most people use the word **"validation"**.

It's time to stop fighting that, and use consistent names.

The word "parse" will no longer be used except when talking about JSON parsing, see
[model methods](#model-namespace-cleanup) above.

## Changes to custom field types :neutral_face:

Since the core structure of validators has changed from "a list of validators to call one after another" to
"a tree of validators which call each other", the 
[`__get_validators__`](https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__)
way of defining custom field types no longer makes sense.

> Instead we'll look for the attribute `__pydantic_schema__` which must be a 

Reviewer 4: This all makes sense, the only thing I wonder about is the word "**schema**", because it's the same word used for JSON Schema, and I'm not sure if that could make some confusion.

Also because both types of "schemas" are mainly a dict with string keys.

Reviewer 5: Since it's a pydantic-core schema and that's unlikely to change, maybe `__pydantic_core_schema__`?

Reviewer 4: Humm, dang it, "naming things", that's an NP-Complete problem :joy: 

As @adriangb says, maybe something that explicitly mentions `core` or `validation` could work and help disambiguate.

Maybe it's not that important, but it's just that I have "suffered" and regretted design choices I've made in the past, e.g. naming two different things with the same name "`app`", and then having to deal with disambiguating that everywhere, and then having to keep the two ambiguous names for backwards compatibility and all that.

Anyway, some word ideas, probably none is perfect, but in case you like any of those:

* definition
* instruction
* declaration
* shape
* design
* architecture
* plan
* pattern
* blueprint
* structure
* arrangement
* prescription
* skeleton
* outline
* frame
* formula

Reviewer 2: We could also rename the `schema` method into `model_json_schema` instead of `model_schema`.
I know there is also the `model_schema_json` but IMHO it could be removed in v2. Don't really see the advantage compared to a simple `(or)json.dumps...` of the json schema as dict

Author: Okay, I've called this thing `__pydantic_validation_schema__`.

@PrettyWood @tiangolo if you can be bothered, it might be worth you looking at the new table I added in "Model Namespace Cleanup" section.

        if 'foobar' in value:
            return Foobar(value)
        else:
            raise ValueError('expected foobar')
```

What's going on here: `__pydantic_schema__` defines a schema which effectively says:

> Validate input data as a string, then call the `validate` function with that string, use the returned value
> as the final result of validation.

`ValidationSchema` is just an alias to 
[`pydantic_core.Schema`](https://github.com/samuelcolvin/pydantic-core/blob/main/pydantic_core/_types.py#L291)
which is a type defining the schema for validation schemas.

!!! note
    pydantic-core schema has full type definitions although since the type is recursive, 
    mypy can't provide static type analysis, pyright however can.

We can probably provide one or more helper functions to make `__pydantic_schema__` easier to generate.

## Other Improvements :thumbsup:

Some other things which will also change, IMHO for the better:

1. Recursive models with cyclic references - although recursive models were supported by pydantic V1,
   data with cyclic references caused recursion errors, in pydantic-core cyclic references are correctly detected
   and a validation error is raised

> 2. The reason I've been so keen to get pydantic-core to compile and run with wasm is that I want all examples

Reviewer 4: I can also imagine (but not sure if crazy) some frontend JS stuff talking directly to the wasm thing, without involving any Python, running the same validation done on backend on the frontend. Probably doesn't make sense, but anyway, would be cool.

Author: Ye, sounds good. This is what @PrettyWood and I joking refer to as "tydantic".

As soon as I get a multiyear prison sentence with access to a computer and internet access.

https://github.com/samuelcolvin/pydantic-core/issues/71

To do it properly, we would:
* build a library with a common interface to pyo3 and rust types, with a compile time flag to switch between pyo3 and rust types
* replace use of pyo3 in pydantic-core with that library
* build js/typescript bindings

Effectively it would be something like

```rust
struct GeneralDict {
  #![cfg(feature = "python")]
  inner: PyDict,
  #![cfg(not(feature = "python"))]
  inner: IndexMap<Value, Value>,
}
```
then outwardly identical interfaces.

  :octicons-clock-24: 10 min read
</div>
</aside>

I've spoken to quite a few people about pydantic V2, and mention it in passing even more.

I owe people a proper explanation of the plan for V2:

* What we will add
* What we will remove
* What we will change
* How I'm intending to go about completing it and getting it released
* Some idea of timeframe :fearful:

Here goes...

## Plan & Timeframe

I'm currently taking a kind of sabbatical after leaving my last job to get pydantic V2 released.
Why? I ask myself that question quite often.
I'm very proud of how much pydantic is used, but I'm less proud of its internals.
Since it's something people seem to care about and use quite a lot 
(26m downloads a month, used by 72k public repos, 10k stars).
I want it to be as good as possible.

While I'm on the subject of why, how and my odd sabbatical: if you work for a large company who use pydantic a lot,
you might encourage the company to **sponsor me a meaningful amount**, 
like [Salesforce did](https://twitter.com/samuel_colvin/status/1501288247670063104).
This is not charity, recruitment or marketing - the argument should be about how much the company will save if

> pydantic is 10x faster, more stable and more powerful - it would be worth paying me 10% of that to make it happen.

Reviewer 4: Good point @adriangb! Yep, totally this. :100: 

In fact, I think the same idea of some type of "contract for pydantic development" could be easier for the people that would want their companies to put money on this but need to justify it to management, finance, and boards in some way. And it doesn't have to really involve anything more than what the open source sponsorship already would, but the naming and legal "semantics" of a contract could make the whole thing a lot easier to handle.

for python.

The motivation for building pydantic-core in Rust is as follows:

1. **Performance**, see [below](#performance)
2. **Recursion and code separation** - with no stack and little or no overhead for extra function calls, 
   Rust allows pydantic-core to be implemented as a tree of small validators which call each other, without harming performance
3. **Safety and complexity** - pydantic-core is a fairly complex piece of code which has to draw distinctions 
   between many different errors, Rust is great in situations like this, 
   it should minimise bugs (:fingers_crossed:) and allow the codebase to be extended for a long time to come

!!! note
    The python interface to pydantic shouldn't change as a result of using pydantic-core, instead
    pydantic will use type annotations to build a schema for pydantic-core to use.

pydantic-core is usable now, albeit with an unintuitive API, if you're interested, please give it a try.

pydantic-core provides validators for all common data types, 
[see a list here](https://github.com/samuelcolvin/pydantic-core/blob/main/pydantic_core/_types.py#L291).

> Other, less commonly used data types will be supported via validator functions implemented in pydantic.

Author: i think this is now covered (I think you were reviewing at the same time as @tiangolo 😖 )

(and significant improvements in how validation objects are structured) pydantic V2 will be significantly faster
than pydantic V1.

Looking at the pydantic-core [benchmarks](https://github.com/samuelcolvin/pydantic-core/tree/main/tests/benchmarks)
today, pydantic V2 is between 4x and 50x faster than pydantic V1.9.1.

In general, pydantic V2 is about 17x faster than V1 when validating a model containing a range
of common fields.

### Strict Mode :thumbsup:

People have long complained about pydantic for coercing data instead of throwing an error.
E.g. input to an `int` field could be `123` or the string `"123"` which would be converted to `123`.

pydantic-core comes with "strict mode" built in. With this only the exact data type is allowed, e.g. passing
`"123"` to an `int` field would result in a validation error.

This will allow pydantic V2 to offer a `strict` switch which can be set on either a model or a field.

#### `is_instance` checks :thumbsup:

Strict mode also means it makes sense to provide an `is_instance` method on validators which effectively run
validation then throws away the result while avoiding the (admittedly small) overhead of creating and raising

> and error or returning the validation result.

Author: makes sense, I'll move it down.

| `int`      | `b"1"`                  | :material-close:      | :material-close: | Error   |

(For the last case converting `bytes` to an `int` could reasonably mean `int(bytes_data.decode())` or 
`int.from_bytes(b'1', 'big/little')`, hence an error)

In addition to the general rule, we'll provide a conversion table which defines exactly what data will be allowed 
to which field types. See [the table below](#conversion-table) for a start on this.

### Built in JSON support :thumbsup:

pydantic-core can parse JSON directly into a model or output type, this both improves performance and avoids
issue with strictness - e.g. if you have a strict model with a `datetime` field, the input must be a 
`datetime` object, but clearly that makes no sense when parsing JSON which has no `datatime` type.
Same with `bytes` and many other types.

Pydantic V2 will therefore allow some conversion when validating JSON directly, even in strict mode 
(e.g. `ISO8601 string -> datetime`, `str -> bytes`) even though this would not be allowed when validating
a python object.

In future direct validation of JSON will also allow:

> * parsing in a separate thread while starting validation in the main thread

Author: Totally possible, I think the best advantage would actually be the ability to stop parsing when you hit the wrong type, without completing all parsing then starting validation.

Since both JSON parsing and validation are pretty big, I imagine the input data would have to be massive before you saw a meaningful difference from streaming parsing vs. reading the whole string of the wire, then starting validation.

  (pydantic-core is [already](https://github.com/samuelcolvin/pydantic-core/runs/7214195252?check_suite_focus=true) 
  compiled for wasm32 using emscripten and unit tests pass, except where cpython itself has 
  [problems](https://github.com/pyodide/pyodide/issues/2841))

Other binaries can be added provided they can be (cross-)compiled on github actions.
If no binary is available from PyPI, pydantic-core can be compiled from source if Rust stable is available.

The only place where I know this will cause problems is Raspberry Pi, which is a 
[mess](https://github.com/piwheels/packages/issues/254) when it comes to packages written in Rust for python.
Effectively, until that's fixed you'll likely have to install pydantic with 
`pip install -i https://pypi.org/simple/ pydantic`.

### Pydantic becomes a pure python package :thumbsup:

Pydantic V1.X is a pure python code base but is compiled with cython to provide some performance improvements.
Since the "hot" code is moved to pydantic-core, pydantic itself can go back to being a pure python package.

This should significantly reduce the size of the pydantic package and make unit tests of pydantic much faster.

> In addition, some constraints on pydantic code can be removed once it no-longer has to be compilable with cython.

Author: very good point, same for reading signatures etc.

   providing validation schema to a `TypeDict` field/item will use `Annotated`, e.g. `Annotated[str, Field(strict=True)]`
4. `from_orm` has become `from_attributes` and is now defined at schema generation time 
   (either via model config or field config)
5. `input_value` has been added to each line error in a `ValidationError`, making errors easier to understand,
   and more comprehensive details of errors to be provided to end users, 
   [pydantic#784](https://github.com/samuelcolvin/pydantic/issues/784)
6. `on_error` logic in a schema which allows either a default value to be used in the event of an error,
   or that value to be omitted (in the case of a `full=False` `TypeDict`),
   [pydantic-core#151](https://github.com/samuelcolvin/pydantic-core/issues/151)
7. `datetime`, `date`, `time` & `timedelta` validation is improved, see the 
   [speedate] Rust library I built specifically for this purpose for more details
8. Powerful "priority" system for optionally merging or overriding config in sub-models for nested schemas
9. pydantic will support [annotated-types](https://github.com/annotated-types/annotated-types), 

>    so you can do stuff like `Annotated[set[int], Len(0, 10)]`

Author: I've added both.

* `dataclass` support, again internals might change, but not the external interface
* `validate_arguments`, might be renamed, but otherwise remain
* hypothesis plugin, might be able to improve this as part of the general cleanup

## Questions :question:

I hope the explanation above is useful. I'm sure people will have questions and feedback; I'm aware
I've skipped over some features with limited detail (this post is already fairly long :sleeping:).

To allow feedback without being overwhelmed, I've created a "Pydantic V2" category for 
[discussions on github](https://github.com/samuelcolvin/pydantic/discussions/categories/pydantic-v2) - please
feel free to create a discussion if you have any questions or suggestions.
We will endeavour to read and respond to everyone.

---

## Implementation Details :nerd:

(This is yet to be built, so these are nascent ideas which might change)

At the center of pydantic v2 will be a `PydanticValidator` class which looks roughly like this:

```py title="PydanticValidator"
# type identifying data which has been validated,
# as per pydantic-core, this can include "fields_set" data
ValidData = TypeVar('ValidData')

> class PydanticValidator:

Author: It's more just pseudo code, I'll add a comment to that effect.

than pydantic V1.

Looking at the pydantic-core [benchmarks](https://github.com/samuelcolvin/pydantic-core/tree/main/tests/benchmarks)
today, pydantic V2 is between 4x and 50x faster than pydantic V1.9.1.

In general, pydantic V2 is about 17x faster than V1 when validating a model containing a range
of common fields.

### Strict Mode :thumbsup:

People have long complained about pydantic for coercing data instead of throwing an error.
E.g. input to an `int` field could be `123` or the string `"123"` which would be converted to `123`.

pydantic-core comes with "strict mode" built in. With this only the exact data type is allowed, e.g. passing
`"123"` to an `int` field would result in a validation error.

This will allow pydantic V2 to offer a `strict` switch which can be set on either a model or a field.

### Formalised Conversion Table :thumbsup:

As well as complaints about coercion, another (legitimate) complaint was inconsistency around data conversion.

In pydantic V2, the following principle will govern when data should be converted in "lax mode" (`strict=False`):

> If the input data has a SINGLE and INTUITIVE representation, in the field's type, AND no data is lost

> > during the conversion, then the data will be converted, Otherwise a validation error is raised.

Reviewer 6: ```suggestion
> during the conversion, then the data will be converted; otherwise a validation error is raised.
```

It can therefore validate a single `string` or `datetime` value, a `TypeDict` or a `Model` equally easily.

This feature will provide significant addition performance improvements in scenarios like:

* Adding validation to `dataclasses`
* Validating URL arguments, query strings, headers, etc. in FastAPI
* Adding validation to `TypedDict`
* Function argument validation
* Adding validation to your custom classes, decorators...

In effect - anywhere where you don't care about a traditional model class instance.

We'll need to add standalone methods for generating JSON Schema and dumping these objects to JSON, etc.

### Required vs. Nullable Cleanup :thumbsup:

Pydantic previously had a somewhat confused idea about "required" vs. "nullable". This mostly resulted from
my misgivings about marking a field as `Optional[int]` but requiring a value to be provided but allowing it to be 
`None`.

In pydantic V2, pydantic will move to match dataclasses, thus:

```py title="Required vs. Nullable"
from pydantic import BaseModel

class Foo(BaseModel):
    f1: str  # required, cannot be None
    f2: str | None  # required, can be None - same as Optional[str] / Union[str, None]

>     f3: str | None = None  # optional, can be None
>     f4: str = 'Foobar'  # optional, but cannot be None  

Reviewer 6: ```suggestion
    f3: str | None = None  # not required, can be None
    f4: str = 'Foobar'  # not required, but cannot be None  
```

I'd rephrase this for extra clarity 🙂  (and I'm super excited about the cleanup!)

   `pip install pydantic[settings]`?
6. The following `Config` properties will be  removed:
   * `fields` - it's very old (it pre-dates `Field`), can be removed
   * `allow_mutation` will be removed, instead `frozen` will be used
   * `error_msg_templates`, it's not properly documented anyway, error messages can be customised with external logic if required
   * `getter_dict` - pydantic-core has hardcoded `from_attributes` logic
   * `json_loads` - again this is hard coded in pydantic-core
   * `json_dumps` - possibly
   * `json_encoders` - see the export "mode" discussion [above](#improvements-to-dumpingserializationexport)
   * `underscore_attrs_are_private` we should just choose a sensible default
   * `smart_union` - all unions are now "smart"

## Features Remaining :neutral_face:

The following features will remain (mostly) changed:

* Generics
* JSONSchema, internally this will need to change a lot, but hopefully the external interface will remain unchanged
* `dataclass` support, again internals might change, but not the external interface
* `validate_arguments`, might be renamed, but otherwise remain

> * hypothesis plugin, might be able to improve this as part of the general cleanup

Reviewer 6: If Pydantic starts using Annotated and the annotated-types constraint objects, then I think most of the features currently provided by the plugin (i.e. handling constrained types) will just come for free 😁

Unsure if you'll need a Hypothesis plugin at all, at that point, though it's probably still useful if you want Hypothesis to e.g. not generate integers that don't fit in an int64... unless of course you auto-add the constraints for that!

        """
        new, check if data is value for the model
        """ # (2)
    @classmethod
    def model_is_instance_json(
        cls,
        data: str | bytes | bytearray,
        *, context=None
    ) -> bool:
        """
        Same as `model_is_instance`, but from JSON
        """
    def model_dump(
        self,
        include: ... = None,
        exclude: ... = None,
        by_alias: bool = False,
        exclude_unset: bool = False,
        exclude_defaults: bool = False,
        exclude_none: bool = False,
        mode: Literal['unchanged', 'dicts', 'json-compliant'] = 'unchanged',
        converter: Callable[[Any], Any] | None = None
    ) -> Any:
        """
        previously `dict()`, as before
        with new `mode` argument
        """
    def model_dump_json(self, ...) -> str:
        """
        previously `json()`, argument as above
        effectively an alias of `json.dump(self.model_dump(..., mode='json'))`
        """
    def model_json_schema(self, ...) -> dict[str, Any]:
        """
        previously `schema()`, arguments roughly as before
        JSON schema as a dict
        """

>     def model_schema_json(self, ...) -> str:

Author: thanks, done.

        previously `parse_obj()`, validate data
        """
    @classmethod
    def model_validate_json(
        cls,
        data: str | bytes | bytearray,
        *,
        context=None
    ) -> Self:
        """
        previously `parse_raw(..., content_type='application/json')`
        validate data from JSON
        """
    @classmethod
    def model_is_instance(cls, data: Any, *, context=None) -> bool: # (2)
        """
        new, check if data is value for the model
        """
    @classmethod
    def model_is_instance_json(
        cls,
        data: str | bytes | bytearray,
        *,
        context=None
    ) -> bool:
        """
        Same as `model_is_instance`, but from JSON
        """
    def model_dump(
        self,
        include: ... = None,
        exclude: ... = None,
        by_alias: bool = False,
        exclude_unset: bool = False,
        exclude_defaults: bool = False,
        exclude_none: bool = False,
        mode: Literal['unchanged', 'dicts', 'json-compliant'] = 'unchanged',
        converter: Callable[[Any], Any] | None = None
    ) -> Any:
        """
        previously `dict()`, as before
        with new `mode` argument
        """

>     def model_dump_json(self, ...) -> str:

Reviewer 2: Like schema I would remove this personally. It helps understand that all methods that have json in them are to handle json inputs. To dump, you just need to use one json package

Author: Let's see, the point is we might be able to generate the JSON while converting. and therefore save a lot of time.

I'll change the comment to "equivalent to" not "alias of".

        """
        previously `json()`, argument as above
        effectively an alias of `json.dump(self.model_dump(..., mode='json'))`
        """
    def model_json_schema(self, ...) -> dict[str, Any]:
        """
        previously `schema()`, arguments roughly as before
        JSON schema as a dict
        """
    def model_update_forward_refs(self) -> None:
        """
        previously `update_forward_refs()`, update forward references
        """

>     def model_copy(self, ...) -> Self:

Reviewer 2: Wouldn't it be better to add `__copy__` and let people use the copy package?

Author: 👍

        """
        previously `copy()`, arguments roughly as before
        copy a model
        """
    @classmethod
    def model_construct(
        self,
        _fields_set: set[str] | None = None,
        **values: Any
    ) -> Self:
        """
        previously `construct()`, arguments roughly as before
        construct a model with no validation
        """
    @classmethod

>     def model_customise_schema(cls, schema: dict[str, Any]) -> dict[str, Any]:

Reviewer 2: I'm sorry for my British friends but since we don't write colour and behaviour I guess we should write customize haha

Author: 🙈 ok...

        cls,
        data: str | bytes | bytearray,
        *,
        context=None
    ) -> bool:
        """
        Same as `model_is_instance`, but from JSON
        """
    def model_dump(
        self,
        include: ... = None,
        exclude: ... = None,
        by_alias: bool = False,
        exclude_unset: bool = False,
        exclude_defaults: bool = False,
        exclude_none: bool = False,
        mode: Literal['unchanged', 'dicts', 'json-compliant'] = 'unchanged',
        converter: Callable[[Any], Any] | None = None
    ) -> Any:
        """
        previously `dict()`, as before
        with new `mode` argument
        """
    def model_dump_json(self, ...) -> str:
        """
        previously `json()`, arguments as above
        effectively equivalent to `json.dump(self.model_dump(..., mode='json'))`,
        but more performant
        """
    def model_json_schema(self, ...) -> dict[str, Any]:
        """
        previously `schema()`, arguments roughly as before
        JSON schema as a dict
        """
    def model_update_forward_refs(self) -> None:
        """
        previously `update_forward_refs()`, update forward references
        """

>     def model_copy(self, ...) -> Self:

Reviewer 2: Glad it's live. You forgot to remove this

Author: Good catch, will do so later.

Reviewer 2: While I'm at it. There are some typos on `TypedDict` and you use the `full` name instead of `total`

> Validate input data as a string, then call the `validate` function with that string, use the returned value
> as the final result of validation.

`ValidationSchema` is just an alias to
[`pydantic_core.Schema`](https://github.com/samuelcolvin/pydantic-core/blob/main/pydantic_core/_types.py#L291)
which is a type defining the schema for validation schemas.

!!! note
    pydantic-core schema has full type definitions although since the type is recursive,
    mypy can't provide static type analysis, pyright however can.

We can probably provide one or more helper functions to make `__pydantic_validation_schema__` easier to generate.

## Other Improvements :thumbsup:

Some other things which will also change, IMHO for the better:

1. Recursive models with cyclic references - although recursive models were supported by pydantic V1,
   data with cyclic references caused recursion errors, in pydantic-core cyclic references are correctly detected
   and a validation error is raised
2. The reason I've been so keen to get pydantic-core to compile and run with wasm is that I want all examples
   in the docs of pydantic V2 to be editable and runnable in the browser

> 3. Full (pun intended) support for `TypedDict`, including `full=False` - e.g. omitted keys,

Reviewer 2: You mean `total=False` I reckon

>    providing validation schema to a `TypeDict` field/item will use `Annotated`, e.g. `Annotated[str, Field(strict=True)]`

Reviewer 2: TypedDict

4. `from_orm` has become `from_attributes` and is now defined at schema generation time
   (either via model config or field config)
5. `input_value` has been added to each line error in a `ValidationError`, making errors easier to understand,
   and more comprehensive details of errors to be provided to end users,
   [pydantic#784](https://github.com/samuelcolvin/pydantic/issues/784)
6. `on_error` logic in a schema which allows either a default value to be used in the event of an error,

>    or that value to be omitted (in the case of a `full=False` `TypeDict`),

Reviewer 2: Same

   [pydantic-core#151](https://github.com/samuelcolvin/pydantic-core/issues/151)
7. `datetime`, `date`, `time` & `timedelta` validation is improved, see the
   [speedate] Rust library I built specifically for this purpose for more details
8. Powerful "priority" system for optionally merging or overriding config in sub-models for nested schemas
9. Pydantic will support [annotated-types](https://github.com/annotated-types/annotated-types),
   so you can do stuff like `Annotated[set[int], Len(0, 10)]` or `Name = Annotated[str, Len(1, 1024)]`

> 10. A single decorator for general usage - we should add a `valdiate` decorator which can be used:

Reviewer 2: validate

   perhaps it should move to a separate package, perhaps installable alongside pydantic with
   `pip install pydantic[settings]`?
8. The following `Config` properties will be removed:
   * `fields` - it's very old (it pre-dates `Field`), can be removed
   * `allow_mutation` will be removed, instead `frozen` will be used
   * `error_msg_templates`, it's not properly documented anyway, error messages can be customized with external logic if required
   * `getter_dict` - pydantic-core has hardcoded `from_attributes` logic
   * `json_loads` - again this is hard coded in pydantic-core
   * `json_dumps` - possibly
   * `json_encoders` - see the export "mode" discussion [above](#improvements-to-dumpingserializationexport)
   * `underscore_attrs_are_private` we should just choose a sensible default
   * `smart_union` - all unions are now "smart"
9. `dict(model)` functionality should be removed, there's a much clearer distinction now that in 2017 when I 
   implemented this between a model and a dict

## Features Remaining :neutral_face:

The following features will remain (mostly) changed:

> * Generics

Reviewer 2: I think this should change and would be a very nice point in the "other improvements" section. Now that we stop using cython, we should be able to have 
```py
class M(Basemodel, Generic [T])...
```
instead of the `GenericModel`

to allow for customising field keys.

### Validation Context :thumbsup:

Pydantic V2 will add a new optional `context` argument to `model_validate` and `model_validate_json`
which will allow you to pass information not available when creating a model to validators.
See [pydantic#1549](https://github.com/samuelcolvin/pydantic/issues/1549) for motivation.

```py title="Context during Validation"
from pydantic import BaseModel, EmailStr, validator

class User(BaseModel):
    email: EmailStr
    home_country: str
    
    @validator('home_country')
    def check_home_country(cls, v, context):
        if v not in context['countries']:
            raise ValueError('invalid country choice')
        return v

async def add_user(post_data: bytes):
    countries = set(await db_connection.fetch_all('select code from country'))
    user = User.model_validate_json(post_data, context={'countries': countries})
    ...
```

!!! note
    We (actually mostly Sebastián :wink:) will have to make some changes to FastAPI to fully leverage `context`
    as we'd need some kind of dependency injection to build context before validation so models can still be passed as

>     arguments to views. I'm sure he'll be game.

Reviewer 4: I'm game! Haha :sweat_smile: 

I think it's gonna be challenging but I'm confident is solvable/doable.

Author: we might need to add another method to models like

```py
class MyModel(BaseModel):
    @classmethod
    async def model_prepare_context(cls, db_conn: Depends(get_db)) -> dict[str, Any:
        """allows custom logic to prepare context"""
```

I guess `model_prepare_context` doesn't even need to on `BaseModel`, you just document in FastAPI that if it's defined, it'll be called with dependency resolution to prepare context?

Up to you, but I'm very happy to add something like this to pydantic if it helps FastAPI.
